### PR TITLE
Adding blob create service to Nimble

### PR DESCRIPTION
**Pull request by huguesBouvier**
Adding blob create service to Nimble

Description
-----------
Adding blob create service to Nimble
Removing separate attribute adding
Improved memory and speed: Reading and writing is now O(1) instead of O(n), n being the numbers of attributes.
Removed arbitrary limit of N attributes.

Checklist:
----------

- [ x] I have tested my changes. No regression in existing tests.
- [ ] My code is Linted.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: vendors/espressif/boards/esp32/ports/ble/nimble/iot_ble_hal_gatt_server.c [hunk 1]
```diff
@@ -1083,11 +752,294 @@ BTStatus_t prvBTSendResponse( uint16_t usConnId,
...
+        }
+        else
+        {
+            pCharacteristics = pvPortCalloc( prvCountCharacteristics( pxService ) + 1, sizeof( struct ble_gatt_chr_def ) );
+
+            if( pCharacteristics == NULL )
+            {
+                configPRINTF( ( "Could not allocate memory for  characteristic array \n" ) );
+                prvCleanupService( pxService, pSvc );
+                xStatus = eBTStatusNoMem;
```
Comment (yuhui-zheng) [inline vendors/espressif/boards/esp32/ports/ble/nimble/iot_ble_hal_gatt_server.c]: Maybe I missed something. I did not find where memory allocation is cleaned up, in the case that --
- prvCopytoESPUUID() succeeded. 
-  pCharacteristics = pvPortCalloc() failed.
The second allocation needs to be cleaned up.

File: vendors/espressif/boards/esp32/ports/ble/nimble/iot_ble_hal_gatt_server.c [hunk 2]
```diff
@@ -1083,11 +752,294 @@ BTStatus_t prvBTSendResponse( uint16_t usConnId,
...
+            {
+                case eBTDbCharacteristic:
+                    /* Allocate memory for UUID and copies it. */
+                    pCharacteristics[ charCount ].uuid = prvCopytoESPUUID( &pxService->pxBLEAttributes[ index ].xCharacteristic.xUuid, NULL );
+                    pCharacteristics[ charCount ].arg = ( void * ) pxService;
+                    pCharacteristics[ charCount ].access_cb = prvGATTCharAccessCb;
+                    pCharacteristics[ charCount ].flags = prvAFRToESPCharPerm( pxService->pxBLEAttributes[ index ].xCharacteristic.xProperties, pxService->pxBLEAttributes[ index ].xCharacteristic.xPermissions );
+
+                    /* Allocate memory for descriptors. */
+                    pCharacteristics[ charCount ].descriptors = pvPortCalloc( prvCountDescriptor( pxService, index + 1 ) + 1, sizeof( struct ble_gatt_dsc_def ) );
```
Comment (yuhui-zheng) [inline vendors/espressif/boards/esp32/ports/ble/nimble/iot_ble_hal_gatt_server.c]: Also this pvPortCalloc(). If allocation fails would previous allocations need to be cleaned up?

File: vendors/espressif/boards/esp32/ports/ble/nimble/iot_ble_hal_gatt_server.c [hunk 3]
```diff
@@ -391,11 +261,19 @@ ble_uuid_t * prvCopytoESPUUID( BTUuid_t * pxUuid )
     switch( pxUuid->ucType )
     {
         case eBTuuidType16:
-            uuid16 = pvPortCalloc( 1, sizeof( ble_uuid16_t ) );
 
-            if( !uuid16 )
+            if( pUuid == NULL )
             {
-                return NULL;
+                uuid16 = pvPortCalloc( 1, sizeof( ble_uuid16_t ) );
```
Comment (ravibhagavandas) [inline vendors/espressif/boards/esp32/ports/ble/nimble/iot_ble_hal_gatt_server.c]: Its better to always allocate the memory for ble_uuid_t outside the function ? That way, caller need not have to remember to free the memory allocated from inside the function.

File: vendors/espressif/boards/esp32/ports/ble/nimble/iot_ble_hal_gatt_server.c [hunk 4]
```diff
@@ -509,63 +403,109 @@ static int prvGATTCharAccessCb( uint16_t conn_handle,
...
+                    gattOffset = handle;
+                }
+
+                pxService->pusHandlesBuffer[ index ] = handle - gattOffset;
+                break;
+            }
+        }
+
+        serviceIndex++;
+    } while( espServices[ serviceIndex ].type != BLE_GATT_SVC_TYPE_END );
```
Comment (ravibhagavandas) [inline vendors/espressif/boards/esp32/ports/ble/nimble/iot_ble_hal_gatt_server.c]: Bound by MAX_SERVICES  as well ?

File: vendors/espressif/boards/esp32/ports/ble/nimble/iot_ble_hal_gatt_server.c [hunk 5]
```diff
@@ -509,63 +403,109 @@ static int prvGATTCharAccessCb( uint16_t conn_handle,
     return rc;
 }
 
+bool setNewHandle( const ble_uuid_t * uuid,
```
Comment (ravibhagavandas) [inline vendors/espressif/boards/esp32/ports/ble/nimble/iot_ble_hal_gatt_server.c]: looks like this is a private function.  Set function type to static and prefix with prv.

File: vendors/espressif/boards/esp32/ports/ble/nimble/iot_ble_hal_gatt_server.c [hunk 6]
```diff
@@ -509,63 +403,109 @@ static int prvGATTCharAccessCb( uint16_t conn_handle,
...
+                    ( void * ) prvCopytoESPUUID( &pxService->pxBLEAttributes[ index ].xCharacteristicDescr.xUuid, ( ble_uuid_t * ) &uuid128 );
+                    break;
+
+                default:
+                    break;
+            }
+
+            /* Complete handle for CCCD. Since no callback are generated for CCCD (because they are added automatically.
+             * The handle of CCCDs is still 0. */
+            if( index + 1 < pxService->xNumberOfAttributes )
```
Comment (ravibhagavandas) [inline vendors/espressif/boards/esp32/ports/ble/nimble/iot_ble_hal_gatt_server.c]: Should this be inside the next if statement ( if UUID matches ) ?

File: vendors/espressif/boards/esp32/ports/ble/nimble/iot_ble_hal_gatt_server.c [hunk 7]
```diff
@@ -1083,11 +752,294 @@ BTStatus_t prvBTSendResponse( uint16_t usConnId,
 }
 
 /*-----------------------------------------------------------*/
+uint16_t prvCountCharacteristics( BTService_t * pxService )
```
Comment (ravibhagavandas) [inline vendors/espressif/boards/esp32/ports/ble/nimble/iot_ble_hal_gatt_server.c]: Static?

File: vendors/espressif/boards/esp32/ports/ble/nimble/iot_ble_hal_gatt_server.c [hunk 8]
```diff
@@ -39,155 +39,23 @@
...
 
 static struct ble_gatt_access_ctxt g_ctxt;
-
-static BTGattEntity_t xGattTable[ GATT_MAX_ENTITIES ];
-
-static uint8_t xGattTableSize, xAttrNum, xCurrentAttrNum;
-
-static uint16_t xLastSvcHandle;
-
+static struct ble_gatt_svc_def espServices[ MAX_SERVICES + 1 ];
```
Comment (ravibhagavandas) [inline vendors/espressif/boards/esp32/ports/ble/nimble/iot_ble_hal_gatt_server.c]: Why do we need an extra element for espServices?

File: vendors/espressif/boards/esp32/ports/ble/nimble/iot_ble_hal_gatt_server.c [hunk 9]
```diff
@@ -1083,11 +752,294 @@ BTStatus_t prvBTSendResponse( uint16_t usConnId,
...
+    /* Fill in service field. After that start creating characteristics and descriptors individually. */
+    if( xStatus == eBTStatusSuccess )
+    {
+        espServices[ serviceCnt + 1 ].type = BLE_GATT_SVC_TYPE_END;
+
+        if( pxService->pxBLEAttributes[ 0 ].xAttributeType == eBTDbPrimaryService )
+        {
+            pSvc->type = BLE_GATT_SVC_TYPE_PRIMARY;
+        }
+        else if( pxService->pxBLEAttributes[ 0 ].xAttributeType == eBTDbPrimaryService )
```
[the remark this answers is not in the document]
Comment (huguesBouvier) [reply vendors/espressif/boards/esp32/ports/ble/nimble/iot_ble_hal_gatt_server.c]: Good catch, fixed!

File: vendors/espressif/boards/esp32/ports/ble/nimble/iot_ble_hal_gatt_server.c (hunk 1 above)
Comment (huguesBouvier) [reply vendors/espressif/boards/esp32/ports/ble/nimble/iot_ble_hal_gatt_server.c]: Good catch too! Fixed

File: vendors/espressif/boards/esp32/ports/ble/nimble/iot_ble_hal_gatt_server.c (hunk 2 above)
Comment (huguesBouvier) [reply vendors/espressif/boards/esp32/ports/ble/nimble/iot_ble_hal_gatt_server.c]: Yes, but this one seems ok. It gets cleaned here: 
            if( xStatus != eBTStatusSuccess )
            {
                configPRINTF( ( "Failed to allocate memory during Attribute creation\n" ) );
                prvCleanupService( pxService, pSvc );
                break;
            }

File: vendors/espressif/boards/esp32/ports/ble/nimble/iot_ble_hal_gatt_server.c (hunk 4 above)
Comment (huguesBouvier) [reply vendors/espressif/boards/esp32/ports/ble/nimble/iot_ble_hal_gatt_server.c]: No, that is why Nimble request the MAX_SERVICE+1.
The last service always has BLE_GATT_SVC_TYPE_END. I had that same comment when I read the code

File: vendors/espressif/boards/esp32/ports/ble/nimble/iot_ble_hal_gatt_server.c (hunk 5 above)
Comment (huguesBouvier) [reply vendors/espressif/boards/esp32/ports/ble/nimble/iot_ble_hal_gatt_server.c]: Thank you! Fixed!

File: vendors/espressif/boards/esp32/ports/ble/nimble/iot_ble_hal_gatt_server.c (hunk 7 above)
Comment (huguesBouvier) [reply vendors/espressif/boards/esp32/ports/ble/nimble/iot_ble_hal_gatt_server.c]: Fixed. Also fixed prvCleanupService

File: vendors/espressif/boards/esp32/ports/ble/nimble/iot_ble_hal_gatt_server.c (hunk 8 above)
Comment (huguesBouvier) [reply vendors/espressif/boards/esp32/ports/ble/nimble/iot_ble_hal_gatt_server.c]: This is required by the Nimble stack. It replaces the length parameters I suppose.

**Pull request merged**